nixos/modules/tasks/filesystems/overlayfs.nix

```diff
@@ -0,0 +1,151 @@
+{ config, lib, pkgs, utils, ... }:
+
+let
+  toplevelConfig = config;
+
+  # Returns a service that creates the required directories before the mount is
+  # created.
+  preMountService = _name: fs:
+    let
+      rwName = "rw-${utils.escapeSystemdPath fs.mountPoint}";
+
+      # The scripted initrd contains some magic to add the prefix to the
+      # paths just in time, so we don't add it here.
+      prefix =
+        if config.boot.initrd.systemd.enable && fs.neededForBoot then
+          "/sysroot"
+        else
+          "";
+
+      upperdir = prefix + fs.overlay.upperdir;
+      workdir = prefix + fs.overlay.workdir;
+    in
+    lib.mkIf (fs.overlay.upperdir != null)
+      {
+        "${rwName}" = {
+          # Wait for sysroot to be mounted before creating upperdir and
+          # workdir.
+          after = [ "sysroot.mount" ];
```

Comment:
This should probably be `RequiresMountsFor=${upperdir} ${workdir}` instead, in case there intermediate mounts.

Reply:
Unresolving because this also raises the question of whether the mounts needed for lower/upper dir should also be marked as `neededForBoot` (they probably should, and I'm not sure if the logic involved there is trivial...)

Reply:
What matters is with RequiresMountsFor=, systemd will do the right thing. If you, however include mounts in your overlay options that do not exist in the intird, there is nothing we can do.

Reply:
I suppose the same problem exists for bind mounts, and it's not introduced by this PR, so let's open a new issue for it.